File: 15/main.go
package main

import (
	"fmt"
	"utils/utils"
)

func speak(val int, spoken map[int][]int) int {
	hist := spoken[val]
	if len(hist) == 1 {
		return 0
	}

	return hist[len(hist)-1] - hist[len(hist)-2]
}

func main() {
	nums := utils.ScanNumbers("big.txt", ",")

	spoken := make(map[int][]int)
	for i, n := range nums {
		spoken[n] = append(spoken[n], i+1)
	}

	last := nums[len(nums)-1]
	for i := len(nums) + 1; i <= 30000000; i++ {
		last = speak(last, spoken)
		spoken[last] = append(spoken[last], i)
		if i == 2020 {
			fmt.Println("\nPart 1: ", last)
		}
	}

	fmt.Println("Part 2: ", last)
}

File: 15/main.cpp
#include <array>

constexpr uint32_t size = 30000000;

int main() {
    std::array<uint64_t, size> seen{};
    std::array nums{7,12,1,0,16,2};
    uint64_t i = 1;
    for ( auto n : nums )
        seen[n] = i++;

    uint32_t last = nums.back();
    for (uint32_t i = nums.size() + 1; i <= size; i++ ) {
        uint64_t hist = seen[last];
        uint32_t high = hist >> 32;
        uint32_t low = uint32_t(hist);
        last = high == 0 ? 0 : low - high;
        seen[last] = (uint64_t(seen[last]) << 32) + i;
    }

    return last;
}
